[GenericSig Builder] Track and canonicalize same-type-to-concrete constraints

DougGregor opened the pull request:
Track each same-type-to-concrete constraint on the potential archetype
against which it was written, and ensure that the requirement sources
for such same-type constraints stay with the potential archetypes on
which they were described. This is similar to the way we track
same-type constraints among potential archetypes.

Use this information to canonicalize same-type-to-concrete constraints
appropriately. For each connected component within an equivalence
class of potential archetypes, select the best requirement source to
the concrete type, or substitute in an abstract requirement source if
none exists. This approach ensures that components that would be
equivalent to that concrete type anyway get a derived source, while
components that get the concrete-type equivalence by being tied to
another component end up with (abstract) explicit requirements to the concrete type.

To get here, we also needed to change the notion of the archetype
anchor so that potential archetypes with no same-type constraints
directly in their path are preferred over potential archetypes that do
have a same-type constraint in their path. Otherwise, the anchor might
be something that is always concrete and is, therefore, not terribly
interesting.

Fixes the new case that popped up in rdar://problem/30478915.

DougGregor committed:
Revert "[Generic signature builder] Eliminate "concrete type in path" checks."

This reverts commit 58f181a470f2a531bdfbb62079df86fafa25a3ef.

DougGregor committed:
Revert "[Generic signature builder] Align the representative with the archetype anchor."

This reverts commit 0df3adb7a37dcd962bdf032c7652dcae2857ceb6.

DougGregor committed:
[GenericSig Builder] Track and canonicalize same-type-to-concrete constraints.

Track each same-type-to-concrete constraint on the potential archetype
against which it was written, and ensure that the requirement sources
for such same-type constraints stay with the potential archetypes on
which they were described. This is similar to the way we track
same-type constraints among potential archetypes.

Use this information to canonicalize same-type-to-concrete constraints
appropriately. For each connected component within an equivalence
class of potential archetypes, select the best requirement source to
the concrete type, or substitute in an abstract requirement source if
none exists. This approach ensures that components that would be
equivalent to that concrete type anyway get a derived source, while
components that get the concrete-type equivalence by being tied to
another

To get here, we also needed to change the notion of the archetype
anchor so that potential archetypes with no same-type constraints
directly in their path are preferred over potential archetypes that do
have a same-type constraint in their path. Otherwise, the anchor might
be something that is always concrete and is, therefore, not terribly
interesting.

Fixes the new case that popped up in rdar://problem/30478915.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Oh, requirement sources. Right.

DougGregor committed:
[GenericSig Builder] Make resolve() less representative-dependent.

`GenericSignatureBuilder::resolve()` was always jumping to the
representative, to treat typealias representatives as
special. However, doing this means that we put the same-type
constraint on the wrong potential archetype (with the wrong source).

Eliminate the use of `getRepresentative()`. This unfortunately
causes us to need recursive resolution, because we can't always depend
on jumping to the representative to avoid it.

DougGregor commented:
@huonw the non-recursive resolution in the new `resolve()` ended up causing some fun issues here. Specifically, the last commit that removes the `getRepresentative()` from that function required me to do recursive expansion (which is a bit of a bummer).

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
@swift-ci please smoke test

DougGregor commented:
Interesting failure on Linux:

    swift::GenericEnvironment::getSubstitutionMap(SubstitutionList) const: Assertion `contextTy->hasError()' failed.

DougGregor commented:
@swift-ci please test OS X

slavapestov commented:
@DougGregor You have a type that appears in getAllDependentTypes(), but maps to a concrete type... this isn't supposed to happen.

DougGregor commented:
@swift-ci please smoke test Linux

huonw commented:
> @huonw the non-recursive resolution in the new resolve() ended up causing some fun issues here. Specifically, the last commit that removes the getRepresentative() from that function required me to do recursive expansion (which is a bit of a bummer).


I think switching it to be recursive should be fine: the only reason it wasn't was that it didn't need to be at the time.

DougGregor commented:
Okay. I actually hope we can avoid the recursion, because in destroying type sugar we lose some information that I don't want to lose, but it Needs More Thought.

DougGregor committed:
[Generic environment] Weaken an assertion when forming a substitution map.

It turns out that GenericSignature::getAllDependentTypes() sometimes
includes generic type parameter types that have been made
concrete. Tolerate this for now, because fixing it causes issues
elsewhere.

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@swift-ci please smoke test and merge

DougGregor commented:
@slavapestov, I weakened the assert. But I still want to look into the generic parameters that are ending up concrete but are still in getAllDepwndentTypes

swiftix commented:
@DougGregor Very nice! Could you also try to fix rdar://30610428? This is the next blocker for the partial specialization.
